jakarta.ls/src/main/java/org/eclipse/lsp4jakarta/JakartaLanguageServer.java

```diff
@@ -70,8 +71,10 @@ public synchronized void updateSettings(Object initializationOptionsSettings) {
 
     @Override
     public CompletableFuture<Object> shutdown() {
-        // TODO Auto-generated method stub
-        return null;
+        // when shutting down LS, TextDocumentService.didClose() may not be called
+        // properly, need to clear existing diagnostics
+        ((JakartaTextDocumentService) textDocumentService).cleanDiagnostics();
```

Comment:
Why do we need to publish diagnostics when we are shutting down?

Reply:
it's not publishing diagnostics, it's cleaning up existing diagnostics.

Reply:
I see that it is called `cleanDiagnostics` but the code it runs is `jakartaLanguageServer.getLanguageClient().publishDiagnostics(new PublishDiagnosticsParams(doc.getUri(), new ArrayList<Diagnostic>())` basically just `publishDiagnostics`. I wasn't able to find the source for `publishDiagnostics`.

Reply:
no, `cleanDiagnostics` is LS private method.

Reply:
not sure what do you mean by "I wasn't able to find the source for publishDiagnostic", that's the API pushing the diagnostic messages to edtior UI, there must be one in IntelliJ, otherwise, you won't see those messages in IntelliJ

Reply:
by the way, can we do this kind of conversation at some other place (e.g. Slack)? instead of holding this PR to be merged.

Reply:
One more thing to be clarify here, `cleanDiagnostics` is the `protected` API specific to `JakartaTextDocumentService`, not defined for generic `TextDocumentService` interface.

Reply:
My point is that if the `JakartaTextDocumentService` needs this ability most likely the MicroProfile language server needs this ability also since they do mostly the same thing. That is why I'm wondering if this ability is needed in the more general interface. Your use case seemed like it could happen with a MicroProfile document as well as this type.

Reply:
I don't know whether MP has this issue or not, but I do see the issue in Jakarta LS, if you concern the differences between MP and Jakarta, probably open up an issue about that and do investigation after MVP?

Reply:
Are you concerning extra "cleanup" steps being called? if you look at the code closely, the `cleanDiagnostics` is just doing a "back-up" step, if there is no message left over, the API will do nothing.